### PR TITLE
(PCP-788) Support file input_method

### PR DESCRIPTION
**Pull request by MikaelSmith**
Supports a `file` input_method that serializes the json input to a file
and passes the name of that file as the 1st argument to the task.

### CODE REVIEW

File: lib/src/modules/task.cc [hunk 1]
```diff
@@ -89,20 +89,32 @@ static const std::string TASK_WRAPPER_EXECUTABLE { "task_wrapper" };
...
     }; }},
-    {".pp",  [](std::string filename) { return TaskCommand {
-        "puppet", { "apply", filename }
+    {".pp",  [](std::string filename, std::string input_file) { return TaskCommand {
+        "puppet", { "apply", filename, input_file }
     }; }},
-    {".ps1", [](std::string filename) { return TaskCommand {
-        "powershell", { "-NoProfile", "-NonInteractive", "-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filename }
+    {".ps1", [](std::string filename, std::string input_file) { return TaskCommand {
+        "powershell", { "-NoProfile", "-NonInteractive", "-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filename, input_file }
```
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/src/modules/task.cc]: From `powershell --help`
```
PowerShell[.exe] [-PSConsoleFile <file> | -Version <version>]
    [-NoLogo] [-NoExit] [-Sta] [-Mta] [-NoProfile] [-NonInteractive] [-Interactive]
    [-InputFormat {Text | XML}] [-OutputFormat {Text | XML}]
    [-WindowStyle <style>] [-EncodedCommand <Base64EncodedCommand>]
    [-ConfigurationName <string>]
    [-Command { - | <script-block> [-args <arg-array>]
                  | <string> [<CommandParameters>] } ]
    [-File <filePath> <args>] [-ExecutionPolicy <ExecutionPolicy>]
```

We match the last one, so yes.

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/src/modules/task.cc]: Windows version
```
> powershell.exe /help

PowerShell[.exe] [-PSConsoleFile <file> | -Version <version>]
    [-NoLogo] [-NoExit] [-Sta] [-Mta] [-NoProfile] [-NonInteractive]
    [-InputFormat {Text | XML}] [-OutputFormat {Text | XML}]
    [-WindowStyle <style>] [-EncodedCommand <Base64EncodedCommand>]
    [-ConfigurationName <string>]
    [-File <filePath> <args>] [-ExecutionPolicy <ExecutionPolicy>]
    [-Command { - | <script-block> [-args <arg-array>]
                  | <string> [<CommandParameters>] } ]
```

File: lib/src/modules/task.cc [hunk 2]
```diff
@@ -89,20 +89,32 @@ static const std::string TASK_WRAPPER_EXECUTABLE { "task_wrapper" };
...
 #endif
 };
 
+// Converts boost filesystem errors to a task_error object.
+static Module::ProcessingError toModuleProcessingError(const fs::filesystem_error& e) {
+    auto err_code = e.code();
+    if (err_code == boost_error::no_such_file_or_directory) {
+        return Module::ProcessingError(lth_loc::format("No such file or directory: {1}", e.path1()));
+    } else if (err_code ==  boost_error::file_exists) {
+        return Module::ProcessingError(lth_loc::format("A file matching the name of the provided sha already exists"));
```
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/src/modules/task.cc]: Ah, probably not. I guess I'll use different handling for that catch block.

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/src/modules/task.cc]: Fixed.

**Comment by MikaelSmith** [comment]
Added acceptance testing for powershell.

File: lib/src/modules/task.cc [hunk 3]
```diff
@@ -445,6 +460,10 @@ ActionResponse Task::callAction(const ActionRequest& request)
                           master_uris_,
                           client_,
                           task_execution_params.get<std::vector<lth_jc::JsonContainer>>("files")));
+    if (!input_file.empty()) {
+        task_command.arguments.push_back(input_file);
```
[the remark this answers is not in the document]
Comment (mruzicka) [reply lib/src/modules/task.cc]: But that's obviously for another PR.

File: lib/src/modules/task.cc [hunk 4]
```diff
@@ -384,7 +393,7 @@ void Task::callNonBlockingAction(
     const std::string &input,
     ActionResponse &response
 ) {
-    const fs::path &results_dir = request.resultsDir();
```
[the remark this answers is not in the document]
Comment (mruzicka) [reply lib/src/modules/task.cc]: Got it, the return value of `ActionRequest::resultsDir()` is a `std::string&`.

File: lib/src/modules/task.cc [hunk 5]
```diff
@@ -435,6 +444,12 @@ ActionResponse Task::callAction(const ActionRequest& request)
         has_input = true;
     }
 
+    if (task_input_method == "file") {
+        input_file = (fs::path(request.resultsDir()) / "input").string();
```
[the remark this answers is not in the document]
Comment (mruzicka) [reply lib/src/modules/task.cc]: I was at that that the `ActionRequest::resultsDir()` returns `fs::path`, which it doesn't so creating it is necessary.

File: lib/src/modules/task.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/src/modules/task.cc]: Right, it's creating a fs::path object. Leaving it a ref would've been fine, just seemed misleading to me.

File: lib/src/modules/task.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply lib/src/modules/task.cc]: Not sure I followed this thought.

**Pull request merged**